### PR TITLE
pump client: some minor update

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
1. add timeout config when create pumps client
2. add function to judge err is retryable
3. some minor fix

### CODE REVIEW

**Comment by WangXiangUSTC** [comment]
@GregoryIan @july2993 PTAL

**Comment by IANTHEREAL** [comment]
can u paste the error u met? I don't know what case is your modification for?

File: tidb-binlog/pump_client/selector.go [hunk 1]
```diff
@@ -212,6 +212,10 @@ func (r *RangeSelector) Next(pump *PumpStatus, binlog *pb.Binlog, retryTime int)
 		return nil
 	}
 
+	if r.Offset >= len(r.Pumps) {
+		r.Offset = 0
+	}
+
 	nextPump := r.Pumps[r.Offset]
```
Comment (WangXiangUSTC) [inline tidb-binlog/pump_client/selector.go:219]: this may cause panic if remove a pump, add a check

File: tidb-binlog/pump_client/client.go [hunk 2]
```diff
@@ -104,7 +105,7 @@ type PumpsClient struct {
 }
 
 // NewPumpsClient returns a PumpsClient.
-func NewPumpsClient(etcdURLs string, securityOpt pd.SecurityOption) (*PumpsClient, error) {
+func NewPumpsClient(etcdURLs string, timeout time.Duration, securityOpt pd.SecurityOption) (*PumpsClient, error) {
```
Comment (WangXiangUSTC) [inline tidb-binlog/pump_client/client.go:105]: tidb has a config to set write binlog timeout, add a argument in this function

File: tidb-binlog/pump_client/client.go [hunk 3]
```diff
@@ -420,9 +422,12 @@ func (c *PumpsClient) Close() {
 	Logger.Infof("[pumps client] is closed")
 }
 
-func isCriticalError(err error) bool {
-	// TODO: add some critical error.
-	return false
+func isRetryableError(err error) bool {
+	if strings.Contains(err.Error(), "received message larger than max") {
```
Comment (IANTHEREAL) [inline tidb-binlog/pump_client/client.go]: what about it?

Comment (WangXiangUSTC) [reply tidb-binlog/pump_client/client.go]: the message size reach the grpc's maxRecvMsgSize, this error is not retryable, return err directly.

File: tidb-binlog/pump_client/client.go [hunk 4]
```diff
@@ -222,7 +223,7 @@ func (c *PumpsClient) WriteBinlog(binlog *pb.Binlog) error {
 		time.Sleep(RetryInterval)
 	}
 
-	return ErrWriteBinlog
+	return err1
```
Comment (WangXiangUSTC) [inline tidb-binlog/pump_client/client.go]: return error make it easy to troubleshoot

File: tidb-binlog/pump_client/client.go (hunk 3 above)
Comment (IANTHEREAL) [reply tidb-binlog/pump_client/client.go]: `code == ResourceExhausted` is not retryable

**Comment by IANTHEREAL** [comment]
Rest LGTM

**Comment by WangXiangUSTC** [comment]
@july2993 PTAL

File: tidb-binlog/pump_client/client.go [hunk 5]
```diff
@@ -420,9 +421,16 @@ func (c *PumpsClient) Close() {
...
 
-func isCriticalError(err error) bool {
-	// TODO: add some critical error.
-	return false
+func isRetryableError(err error) bool {
+	// ResourceExhausted is a error code in grpc.
+	// ResourceExhausted indicates some resource has been exhausted, perhaps
+	// a per-user quota, or perhaps the entire file system is out of space.
+	// https://github.com/grpc/grpc-go/blob/9cc4fdbde2304827ffdbc7896f49db40c5536600/codes/codes.go#L76
+	if strings.Contains(err.Error(), "ResourceExhausted") {
```
Comment (july2993) [inline tidb-binlog/pump_client/client.go:426]: why this is not retryable

File: tidb-binlog/pump_client/client.go [hunk 6]
```diff
@@ -194,6 +192,7 @@ func (c *PumpsClient) WriteBinlog(binlog *pb.Binlog) error {
 	}
 	req := &pb.WriteBinlogReq{ClusterID: c.ClusterID, Payload: commitData}
 
+	var err1 error
```
Comment (july2993) [inline tidb-binlog/pump_client/client.go]: you don't need this, just use err

**Review by july2993 — COMMENTED** [review]
you should better fix don't write the C binlog to not the according pump, or it's easy to loss the C binlog, it's meaningless to write to other pump
rest LGTM

File: tidb-binlog/pump_client/client.go (hunk 5 above)
Comment (WangXiangUSTC) [reply tidb-binlog/pump_client/client.go:426]: almost all the errer with ResourceExhausted is `received message larger than max`, it is no need to retry

**Comment by WangXiangUSTC** [comment]
@july2993  I will fix it next pr

**Comment by july2993** [comment]
LGTM

**Pull request merged**